tests/test_cuda_bd_exchange_mover.py

```diff
@@ -0,0 +1,192 @@
+    batch_moved_coords, _ = bdem_b.move(conf, box)
+    # Moves should be deterministic regardless the number of steps taken per move
+    np.testing.assert_array_equal(iterative_moved_coords, batch_moved_coords)
+
+
+@pytest.mark.parametrize("steps_per_move", [1, 10])
+@pytest.mark.parametrize("moves", [2500])
+@pytest.mark.parametrize("precision,rtol,atol", [(np.float64, 5e-6, 5e-6), (np.float32, 1e-4, 2e-3)])
+@pytest.mark.parametrize("seed", [2023])
+def test_moves_in_a_water_box(steps_per_move, moves, precision, rtol, atol, seed):
+    """Verify that the log acceptance probability between the reference and cuda implementation agree"""
+    ff = Forcefield.load_default()
+    system, conf, box, _ = builders.build_water_system(3.0, ff.water_ff)
+    bps, _ = openmm_deserializer.deserialize_system(system, cutoff=1.2)
+
+    nb = next(bp for bp in bps if isinstance(bp.potential, Nonbonded))
+    bond_pot = next(bp for bp in bps if isinstance(bp.potential, HarmonicBond)).potential
+
+    group_idxs = get_group_indices(get_bond_list(bond_pot), conf.shape[0])
+
+    N = conf.shape[0]
+
+    params = nb.params
+
+    cutoff = nb.potential.cutoff
+    klass = custom_ops.BDExchangeMove_f32
+    if precision == np.float64:
+        klass = custom_ops.BDExchangeMove_f64
+
+    bdem = klass(N, group_idxs, params, DEFAULT_TEMP, nb.potential.beta, cutoff, seed, steps_per_move)
+
+    ref_bdem = RefBDExchangeMove(nb.potential.beta, cutoff, params, group_idxs, DEFAULT_TEMP)
+
+    assert bdem.last_log_probability() == 0.0, "First log probability expected to be zero"
+    accepted = 0
+    last_conf = conf
+    for _ in range(moves // steps_per_move):
+        x_move, x_box = bdem.move(last_conf, box)
+        # The box will never change
+        np.testing.assert_array_equal(box, x_box)
+        num_moved = 0
+        for i, mol_idxs in enumerate(group_idxs):
+            if not np.all(x_move[mol_idxs] == last_conf[mol_idxs]):
+                num_moved += 1
+        if num_moved > 0:
+            accepted += 1
+            # Verify that the probabilities and per mol energies agree when we do accept moves
+            # can only be done when we only attempt a single move per step
+            if steps_per_move == 1:
+                before_log_weights = ref_bdem.batch_log_weights(last_conf, box)
+                after_log_weights = ref_bdem.batch_log_weights(x_move, x_box)
+                np.testing.assert_allclose(
+                    np.exp(bdem.last_log_probability()),
+                    np.exp(np.minimum(logsumexp(before_log_weights) - logsumexp(after_log_weights), 0.0)),
+                    rtol=rtol,
+                    atol=atol,
+                )
+        if num_moved == 0:
+            np.testing.assert_array_equal(last_conf, x_move)
+        assert num_moved <= 1, "More than one mol moved, something is wrong"
+        last_conf = x_move
+    assert accepted > 0, "No moves were made, nothing was tested"
```

Comment:
what's typical accepted here?

Reply:
(if acceptance rate is someting like 0.01 then 2500*0.01 should be around 25 moves accepted on average?)

Reply:
The acceptance rates I stated before (https://github.com/proteneer/timemachine/pull/1202#issuecomment-1797201324) were percentages, so it is 0.01 **%** which means the expected here would be 0.25 moves. The typical accepted here is 1/2 (float64 results in 1 move, float32 results in 2 moves)

Reply:
oof

Reply:
would this test be a bit brittle? may run 10x longer so we average 10-20 moves?

Reply:
This test is brittle, was a trade off between test time and verifying logic. Since the acceptance rate is so low didn't want to go too longer, especially with how slow it is currently.

Tried 20k steps and only get 5 moves. Trying to find the right combination for a robust nightly test

Reply:
Added a nightly test: https://github.com/proteneer/timemachine/pull/1202/commits/34d84e7a171b1920f6317bd8155742b931f50b2c